tests/test_bookmarks.py

```diff
@@ -0,0 +1,159 @@
+
+                    # Verify the 2nd bookmark is equal to 1st sync bookmark
+                    self.assertEqual(first_bookmark_value, second_bookmark_value)
+
+                    for record in first_sync_messages:
+                        replication_key_value = record.get(replication_key)
+                        # Verify 1st sync bookmark value is the max replication key value for a given stream
+                        self.assertLessEqual(
+                            replication_key_value,
+                            first_bookmark_value,
+                            msg="First sync bookmark was set incorrectly, a record with a greater replication key value was synced",
+                        )
+
+                    for record in second_sync_messages:
+                        replication_key_value = record.get(replication_key)
+                        # Verify the 2nd sync replication key value is greater or equal to the 1st sync bookmarks
+                        self.assertGreaterEqual(
+                            replication_key_value,
+                            simulated_bookmark,
+                            msg="Second sync records do not respect the previous bookmark",
+                        )
+                        # Verify the 2nd sync bookmark value is the max replication key value for a given stream
+                        self.assertLessEqual(
+                            replication_key_value,
+                            second_bookmark_value,
+                            msg="Second sync bookmark was set incorrectly, a record with a greater replication key value was synced",
+                        )
+
+                    # Verify that we get less data in the 2nd sync
+                    self.assertLessEqual(
+                        second_sync_count,
+                        first_sync_count,
+                        msg="Second sync does not have less records, bookmark usage not verified",
+                    )
+
+                elif expected_replication_method == self.FULL_TABLE:
+
+                    # Verify the syncs do not set a bookmark for full table streams
+                    self.assertIsNone(first_bookmark_value)
+                    self.assertIsNone(second_bookmark_value)
+
+                    # Verify the number of records in the second sync is the same as the first
+                    self.assertEqual(second_sync_count, first_sync_count)
+
+                else:
+                    raise NotImplementedError(
+                        "INVALID EXPECTATIONS\t\tSTREAM: {} REPLICATION_METHOD: {}".format(
+                            stream, expected_replication_method
+                        )
+                    )
+
+                # Verify at least 1 record was replicated in the second sync
+                self.assertGreater(second_sync_count, 0, msg=f"We are not fully testing bookmarking for {stream}")
```

Comment:
If this assertion doesn't pass then test will fail #82 with KeyError, so move this above.

Reply:
Moved it